### ISSUE TITLE
Remove the short-cut methods from the Sharer class

### ISSUE THREAD

**Issue by jorabin** [derecalliance/api-java#13]
Bringing forward this conversation:

https://github.com/derecalliance/api-java/pull/12#issuecomment-1820996269

we should consider removing methods, so as to leave only `createSecret` and leave it to the user to update content and add users, without the short-cut methods that do it all in one go

**Comment by ehanoc** [comment]
@jorabin @lbaird Unless you want to try to use optional parameters in Java, which are underwhelming compared to other languages, having a single method with a multiple of run-time path can lead to a confusing interface. 

Since most parameters are part of `DeRecSecret` anyway,  maybe we could apply something like a [Builder Pattern](https://en.wikipedia.org/wiki/Builder_pattern) to `DeRecSecret`. And a single method could look like 

```java
    DeRecSecret newSecret(DeRecSecret secret, boolean recovery);
```

I'm happy to make the contribution if you think it's a good idea

**Comment by jorabin** [comment]
@ehanoc - agreed, the overloaded methods were turning from a "convenience" to being just confusing, which was the motivation behind raising this issue. 

In our (The Building Blocks) prototype implementation we use a builder subclass of our implementation of DeRecSecret rather like what you suggest, though with the primary concern of setting the secret's operational parameters (retry, etc.) whose setting is not standardised in the interface. 

I'm not sure that standardising a builder or the setting of the various operational parameters is something the interface needs to do as different implementations will have different use cases in mind and requiring implementation of irrelevant things for the sake of interface conformance could be seen as unhelpful. 

These parameters should be settable only at secret creation time, so I don't think they should appear in the standard interface, but need to be settable somehow in an actual implementation. While I am personally a big fan of builders, I guess not everyone is, so requiring that people do it that way rather than by their choice of alternative would be to take too strongly an opinionated view?

So, in summary, I think that just simplifying the overloaded methods of the interface to provide for creation of a secret by a sharer would be the right way to go. The `newSecret` method should be a method of `Sharer` to bind the secret to the sharer. I'd be in favour of a reference implementation of the interface providing a builder at least for those operational parameters that need to be set at secret creation time.

**Comment by Loedn** [comment]
This approach should be adopted across all repo's and interfaces, broadening the scope of this issue
